[functorch] introduce an experimental map() op.

zhxchen17 opened the pull request:
Summary:
We want to introduce an experimental control flow op: map() to export some models as FX graphs correctly.

Some calrification on basic requirements we have in mind:
1. This op can nest cond() and other control flow primitives internally.
2. We don't necessarily need loop carried dependencies for the models we've seen.
3. This map() op can handle dynamically shaped tensor as input and return dynamically shaped output based on input shapes.
4. We should be able to pass through additional arguments to the loop body as extra arguments.

In this diff we introduce a new control flow op `map()` which has the following semantics:
```
def map(f: Callable, xs: Tensor, *args):
    # one possible implementation:
    return torch.stack([f(x, *args) for x in xs])
```

Test Plan:
pytest functorch/test_control_flow.py
CI

Differential Revision: D41165796

ezyang commented:
What exactly is being looped over here? Are you permitting lists of varying sizes in the IR format? (If so, that makes me very negative on this feature. cond is OK because it doesn't introduce new data types to the FX IR. This would.)

zhxchen17 commented:
> What exactly is being looped over here? Are you permitting lists of varying sizes in the IR format? (If so, that makes me very negative on this feature. cond is OK because it doesn't introduce new data types to the FX IR. This would.)

@ezyang 

> What exactly is being looped over here?

I think the semantics is very similar to how jax.lax.map works, basically we loop over the second argument:
```
def map(f, xs, *args):
    return torch.stack([f(x, *args) for x in xs])
```

> Are you permitting lists of varying sizes in the IR format?

No we don't permit that, varying size list is not useful to us and we don't want to add it. `xs` argument must be a tensor, and we have a check during tracing:
```
# map.py, line 29
assert isinstance(xs, torch.Tensor), "map() must operate over a tensor"
```

ezyang commented:
OK, makes more sense.

ezyang commented:
deferring mostly to @voznesenskym and @zou3519 for review on this. Holler if you want me to look at something specific.

zou3519 commented on functorch/experimental/cond.py:
```diff
@@ -115,6 +123,17 @@ def inner(pred, true_fn, false_fn, operands):
     return res
 
 
+@cond.py_impl(FakeTensorMode)
+def cond_fake_tensor_mode(pred, true_fn, false_fn, operands):
+    return true_fn(*operands)
```
This is a bit fishy, I'd expect it to run true_fn(*operands), false_fn(*operands), and then assert that everything is identical

zou3519 commented on functorch/experimental/cond.py:
```diff
@@ -115,6 +123,17 @@ def inner(pred, true_fn, false_fn, operands):
+    return true_fn(*operands)
+
+
+@cond.py_impl(DispatchKey.PythonDispatcher)
+def cond_python_dispatcher(*args):
+    _ = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.PythonDispatcher))
+    return cond(*args)
```
nit: this is just cond.fallthrough(DispatchKey.PythonDispatcher), ditto for the map_python_dispatcher

zhxchen17 replied:
@zou3519 I tried to use `fallthrough()` directly, but seems like the current behavior of `fallthrough()` will always add `PythonDispatcher` key back which causes infinite loop: 
```
    def _fallthrough_fn(self, operator, dispatch_key):
        def inner(*args, **kwargs):
            all_keys_after_current = torch._C._dispatch_keyset_full_after(dispatch_key)
            ...
```
and 
```
  constexpr DispatchKeySet(FullAfter, DispatchKey t)
      // LSB after t are OK, but not t itself.
      // "functionalities" have a notion of ordering (e.g. Autograd > Sparse >
      // Quantized > Dense). But backends don't really have an ordering.
      // Therefore, we're enforcing that FullAfter can only be used on
      // "functionality" keys.
      : repr_(
            (1ULL
             << (num_backends + static_cast<uint8_t>(toFunctionalityKey(t)) -
                 1)) -
            1) {
    *this = add(DispatchKey::PythonDispatcher);
  }
```
not sure what's the optimal thing to do here other than just masking the PythonDispatcher key and go.

ezyang replied:
can you file an issue about this. iirc there is a problem with fallthrough interaction and python dispatcher

zhxchen17 replied:
@ezyang filed: https://github.com/pytorch/pytorch/issues/89037

zou3519 replied:
If you want to get this PR shipped to try stuff, no need to wait until that issue is fixed though we should figure it out at some point

zhxchen17 replied:
@zou3519 understood, I'll make a comment here pointing to the issue number.

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -345,5 +346,108 @@ def f(x, y):
+        x = torch.randn(3, 2, 2)
+        y = torch.randn(2)
+        res = gm(x, y)
+        self.assertTrue(torch.allclose(res, g(x, y)))
```
nit: use self.assertEqual, saves a couple of characters

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -345,5 +346,108 @@ def f(x, y):
+        res = gm(x, y)
+        self.assertTrue(torch.allclose(res, g(x, y)))
+
+    def test_map_symbolic(self):
+        def f(x, y):
+            return x + y
+
+        def g(xs, y):
+            return functorch.experimental.map(f, xs, y)
+
+        gm = make_fx(g, tracing_mode="symbolic")(torch.ones(3, 2, 4), torch.ones(4))
+        x = torch.randn(3, 2, 2)
+        y = torch.randn(2)
+        res = gm(x, y)
+        self.assertTrue(torch.allclose(res, g(x, y)))
```
What does the traced graph look like? Can we do an expect test against it?

zou3519 commented on functorch/experimental/__init__.py:
```diff
@@ -1,5 +1,6 @@
+from .._src.eager_transforms import hessian, jacfwd, jvp
 from .._src.vmap import chunk_vmap
+from .batch_norm_replacement import replace_all_batch_norm_modules_
+from .map import map
```
meta-point on API: (1) we should be consistent for the control flow ops, (2) we should avoid exposing implementation details publicly (functorch.experimental.map is technically public API, we will be sad if people start relying on functorch.experimental.map.trace_map for example).

For (1) consistency: cond is functorch.experimental.cond.cond. map is functorch.experimental.map. We are likely going to add some more control flow ops in the future. Let's try to gather all the APIs into a functorch.experimental.control_flow namespace.

For (2) private APIs: the implementation of map should go into functorch.experimental._map.py or functorch/_src/map.py and then a single public API should be exposed in functorch.experimental.control_flow. We didn't do this with cond, but it is good practice.

zhxchen17 replied:
good point, I was wondering whether experimental sounds "private" enough. I can make the adjustments proposed here.

zou3519 commented on functorch/experimental/map.py:
```diff
@@ -0,0 +1,97 @@
+            return e
+        return get_proxy_slot(e, proxy_mode.tracer, e, lambda e: e.proxy)
+
+    assert isinstance(xs, torch.Tensor), "map() must loop over a tensor"
+    assert len(xs.shape) > 0 and xs.shape[0] > 0, "map() cannot be traced with scalar tensors or zero dimension tensors"
+    assert all(isinstance(o, (torch.Tensor, torch.nn.Module)) for o in args), "map() operands must be a list of tensors of modules"
```
nit: should be RuntimeError instead of assertion, these are user-facing errors, right?

zhxchen17 replied:
sure

zou3519 commented on functorch/experimental/map.py:
```diff
@@ -0,0 +1,97 @@
+    out_proxy = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, {},
+                                               name="map")
+    outs = [body_graph(x, *args) for x in xs]
+    out = xs.new_empty([xs.shape[0], *outs[0].shape])
+    out.copy_(torch.stack(outs))
```
why do we need to construct `out` this way? Why couldn't it have been `out = torch.stack(outs)`?

zhxchen17 replied:
@zou3519 this is good question and the reason I didn't use `torch.stack()` is because `torch.stack([...])` will take a Python list which doesn't have the dynamic shape information we want to preserve. For example:
```
xs: Tensor[s0, s1]
l = [x for x in xs]
ys = torch.stack(l)
# ys will be specialized to: Tensor[c, s1] where c is the 
```
ideally we want `ys` to preserve dynamic shape dimensions, e.g. `Tensor[s0, s1]` here as well.

maybe there could be a better way but currently I only got `new_empty()` + `copy_` to work.

zou3519 replied:
I see the problem...

zou3519 replied:
Okay, this seems reasonable to me. It's not like we're actually executing new_empty and then copy_ (which would do one additional tensor allocation that is not necessary), and at this point the only thing we have Tensors (and not Tensor subclasses, where the copy_ would have been problematic).

Could we add a comment in the code about this? (otherwise, someone might do the rewrite to torch.stack and break something)

zhxchen17 replied:
ok np

zou3519 reviewed (commented):
I have some questions, but this seems reasonable to me, it's in line with what we did with cond

zhxchen17 committed:
[functorch] introduce an experimental map() op.

Summary:

We want to introduce an experimental control flow op: map() to export some models as FX graphs correctly.

Some calrification on basic requirements we have in mind:
1. This op can nest cond() and other control flow primitives internally.
2. We don't necessarily need loop carried dependencies for the models we've seen.
3. This map() op can handle dynamically shaped tensor as input and return dynamically shaped output based on input shapes.
4. We should be able to pass through additional arguments to the loop body as extra arguments.

In this diff we introduce a new control flow op `map()` which has the following semantics:
```
def map(f: Callable, xs: Tensor, *args):
    # one possible implementation:
    return torch.stack([f(x, *args) for x in xs])
```

Test Plan:
pytest functorch/test_control_flow.py
CI

Reviewers:

Subscribers:

Tasks:

Tags:

zhxchen17 commented:
addressed all comments.

zhxchen17 commented:
@pytorchbot merge
